File: Aula05/Go/Composicao.go
package main

import "fmt"

type Motor struct {
	Tipo     string
	Potencia int
}

func (m Motor) Ligar() {
	fmt.Println("Motor ligado.")
}

func (m Motor) Desligar() {
	fmt.Println("Motor desligado.")
}

type Pneu struct {
	Marca   string
	Tamanho int
}

func (p Pneu) Inflar() {
	fmt.Println("Pneu inflado.")
}

func (p Pneu) Desinflar() {
	fmt.Println("Pneu desinflado.")
}

type Carro struct {
	Marca  string
	Modelo string
	Motor  Motor
	Pneus  [4]Pneu
}

func CriarCarro(marca, modelo string) Carro {
	motor := Motor{"Gasolina", 160}
	pneus := [4]Pneu{}
	for i := range pneus {
		pneus[i] = Pneu{"Pirelli", 18}
	}
	return Carro{Marca: marca, Modelo: modelo, Motor: motor, Pneus: pneus}
}

func (c Carro) Ligar() {
	c.Motor.Ligar()
	fmt.Println("carro ligado.")
}

func (c Carro) Desligar() {
	c.Motor.Desligar()
	fmt.Println("Carro desligado.")
}

func (c *Carro) SubstituirPneu(pos int, novoPneu Pneu) {
	if pos >= 0 && pos < 4 {
		c.Pneus[pos] = novoPneu
		fmt.Printf("Pneu %d substituído.\n", pos+1)
	} else {
		fmt.Println("Posição de pneu inválida.")
	}
}

func main() {
	carro := CriarCarro("Mitsubishi", "Lancer")
	carro.Ligar()

	novoPneu := Pneu{"Michelin", 17}
	carro.SubstituirPneu(2, novoPneu)

	carro.Desligar()
}

File: Aula05/Python/Composicao.py
class Motor:
    def __init__(self, tipo, potencia):
        self.tipo = tipo
        self.potencia = potencia

    def ligar(self):
        print("O motor esta ligado.")
    def desligar(self):
        print("O motor esta desligado.")

class Pneu:
    def __init__(self, marca, tamanho):
        self.marca = marca
        self.tamanho = tamanho

    def inflar(self):
        print("O pneu esta inflado.")
    def desinflar(self):
        print("O pneu esta desinflado.")

class Carro:
    def __init__(self, marca, modelo):
        self.marca = marca
        self.modelo = modelo
        self.motor = Motor("Gasolina", 160)
        self.pneus = [Pneu("Pirelli", 18) for _ in range(4)]


    def ligar(self):
        self.motor.ligar()
        print("O carro esta ligado.")
    def desligar(self):
        self.motor.desligar()
        print("O carro esta desligado.")
    def trocar_pneu(self, indice, novo_pneu):
        self.pneus[indice] = novo_pneu
        print(f"Pneu {indice + 1} trocado.")

# Exemplo de uso
carro = Carro("Mitsubishi", "Lancer")
carro.ligar()

novo_pneu = Pneu("Michelin", 17)
carro.trocar_pneu(2, novo_pneu)

carro.desligar()
